lib/stats/stats-query.c

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright (c) 2002-2017 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "stats-cluster.h"
+#include "stats-query.h"
+#include "stats-registry.h"
+#include "stats.h"
+
+#include <string.h>
+#include <stdio.h>
+
+static const gchar COUNTER_TEMPLATE[] = "%s.%s\n";
+static const gchar COUNTER_WITH_VALUE_TEMPLATE[] = "%s.%s: %d\n";
+static const gchar *_counter_type[SC_TYPE_MAX+1] =
+{
+  "dropped",
+  "processed",
+  "stored",
+  "suppressed",
+  "stamp",
+  "MAX"
+};
+
+
+typedef gboolean (*StatsClusterCounterCb)(StatsCluster *sc, StatsCounterItem *ctr, const gchar *ctr_name,
+    gpointer user_data);
+
+static void
+_foreach_live_counters(StatsCluster *sc, StatsClusterCounterCb counter_cb, gpointer user_data)
+{
+  for (gint i = 0; i < SC_TYPE_MAX; i++)
+    {
+      if (stats_cluster_is_alive(sc, i))
+        if (!counter_cb(sc, &sc->counters[i], _counter_type[i], user_data))
+          break;
+    }
+}
+
+static gboolean
+_append_counter_with_value(StatsCluster *sc, StatsCounterItem *ctr, const gchar *ctr_name, gpointer user_data)
+{
+  GString *str = (GString *)user_data;
+  g_string_append_printf(str, COUNTER_WITH_VALUE_TEMPLATE, sc->query_key, ctr_name ? ctr_name : "", ctr->value);
+  return TRUE;
+}
+
+static gboolean
+_append_counter_without_value(StatsCluster *sc, StatsCounterItem *ctr, const gchar *ctr_name, gpointer user_data)
+{
+  GString *str = (GString *)user_data;
+  g_string_append_printf(str, COUNTER_TEMPLATE, sc->query_key, ctr_name ? ctr_name : "");
+  return TRUE;
```

Comment:
In this cases I find it more problematic to extract the name of the template as it is tightly coupled with the printf arguments anyway. And when I read the code, I need to break out and locate the template in the top of the file. Why is that useful?